### PR TITLE
add tilt/angle

### PR DESCRIPTION
**Pull request by thesaint1975**
Hi Cornim,

as discussed earlier that day, here is my pull request.
tilt can be adjusted from 30 (open) and 100 (fully closed)
shades[1].set_shade_tilt_position(100)

with shades[1].get_shade_state(True) every of my 6 raffstores (venetian blinds) shows slightly different tilt values, so this are the "lowest common denominator", and the difference in tilt/ angle is imperceptible.

shades[1].stop_moving_shade() stops surprisingly a moving shade, sometimes it needs some seconds bust most of my trials it stops during 1or 2 seconds

please dont hesitate to reject if the quality is insufficient, my coding skills are very limited (as you have surely noticed).

if you have any pointers or clues regarding the HA Integration, it would be really helpful

cheers
roland

### CODE REVIEW

File: warema_wms/shade.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 
 class Shade:
     def __init__(self, wms_ctrl: WmsController, room, channel,
-                 time_between_cmds=0.1, num_retries=3, position=0, is_moving=False):
+                 time_between_cmds=0.1, num_retries=3, position=0, tilt_position=0, is_moving=False):
```
Comment (cornim) [inline warema_wms/shade.py]:
```suggestion
                 time_between_cmds=0.1, num_retries=3, position=0, tilt_position=30, is_moving=False):
```

If vaild values range between 30-100 the default should also be between those values.

File: warema_wms/shade.py [hunk 2]
```diff
@@ -28,6 +28,7 @@ def __init__(self, wms_ctrl: WmsController, room, channel,
         self.time_between_cmds = time_between_cmds
         self.num_retries = num_retries
         self.position = position  # 0 for open 100 for closed
+        self.tilt_position = tilt_position    # lets see what happens now # 30=tilt open 100=tilt closed, 6 Raffstores show different values here, from 20 - 28 & 101 - 104
```
Comment (cornim) [inline warema_wms/shade.py]:
```suggestion
        self.tilt_position = tilt_position    # 30=tilt open 100=tilt closed, 6 Raffstores show different values here, from 20 - 28 & 101 - 104
```

I would limit the comments to technical content

File: warema_wms/shade.py [hunk 3]
```diff
@@ -88,6 +90,40 @@ def set_shade_position(self, new_position):
                        .format(self.room.name, self.channel.name, new_position))
         return False
 
+    def set_shade_tilt_position(self, new_tilt_position):
+        """
+        Sets shade to new_tilt.
+        :param new_tilt: New tilt of shade
```
Comment (cornim) [inline warema_wms/shade.py]:
```suggestion
        :param new_tilt: New tilt of shade (30=open, 100=closed)
```

File: warema_wms/shade.py [hunk 4]
```diff
@@ -88,6 +90,40 @@ def set_shade_position(self, new_position):
...
+        """
+        for _ in range(self.num_retries):
+            self._try_cmd_n_times(lambda: self.wms_ctrl.send_rx_check_ready(self.room.id, self.channel.id),
+                                  self.num_retries)
+            time.sleep(self.time_between_cmds)
+            self.wms_ctrl.send_tx_tilt_shade(self.room.id, self.channel.id, new_tilt_position * 2)
+            # This cmd is sent by the JS app of the web control server but its purpose is unclear and the feedback
+            # is always 0
+            # self.wms_ctrl.send_rx_move_shutter(self.room.id, self.channel.id)
```
Comment (cornim) [inline warema_wms/shade.py]:
```suggestion
```

I would delete that comment. Having it once is probably sufficient.

File: warema_wms/shade.py [hunk 5]
```diff
@@ -105,6 +141,24 @@ def _verify_set_cmd_sent(self, target_position):
             time.sleep(self.time_between_cmds)
         return False
 
+    def _verify_tilt_cmd_sent(self, target_tilt_position):
+        time.sleep(self.time_between_cmds)
+        for _ in range(self.num_retries):
+            self.update_shade_state()
+            if self.is_moving or self.tilt_position == target_tilt_position:
```
Comment (cornim) [inline warema_wms/shade.py:145]: I don't know if `is_moving` is `True` when only the tilt is being changed. Probably something that should be verified.

File: warema_wms/wms_controller.py [hunk 6]
```diff
@@ -18,16 +18,17 @@
 RX_SHADE_STATE = '0431'
 
 TX_MOVE_SHADE = '0821'
-SHADE_POSITION = '03{}ffffff'
+SHADE_POSITION = '03{}dfffff'   # df = trial to move raffstores in tilt closed pos., working in console but not in HA
```
Comment (cornim) [inline warema_wms/wms_controller.py]:
```suggestion
SHADE_POSITION = '03{}ffffff'
```

You probably forgot to change this back after your trail.

File: warema_wms/wms_controller.py [hunk 7]
```diff
@@ -18,16 +18,17 @@
...
+TILT_POSITION = '03ff{}ffff'
```
Comment (cornim) [inline warema_wms/wms_controller.py:22]: This is surprising to me. I would have guessed that position and tilt should be send in one command. The way you set it up is with two commands. One for position where the tilt is always `ff` and one for tilt where the position is always `ff`.
If that's how the warema webcontrol does it, it's probably correct, but as I said, I have no way of verifying it.

File: warema_wms/wms_controller.py [hunk 8]
```diff
@@ -18,16 +18,17 @@
...
+STOP_SHADE = '01ffffffff'       # at least at my setup it works (instead of STOP_SHADE = '0001ffffff')
```
Comment (cornim) [inline warema_wms/wms_controller.py:23]:
```suggestion
STOP_SHADE = '01ffffffff'
```

I believe you. I never actually tried the stop feature out, since I don't need it.

File: warema_wms/wms_controller.py [hunk 9]
```diff
@@ -18,16 +18,17 @@
...
 
 logger = logging.getLogger('warema_wms')
-
+logging.debug
```
Comment (cornim) [inline warema_wms/wms_controller.py]:
```suggestion
```

This should probably not be here.

File: warema_wms/wms_controller.py [hunk 10]
```diff
@@ -69,6 +70,7 @@ def _increment(self):
         self.initial_ts += 1
         return res
 
+
```
Comment (cornim) [inline warema_wms/wms_controller.py]:
```suggestion
```

PEP8 says: Only one empty line between methods.

**Review by cornim — CHANGES_REQUESTED** [review]
Hi Roland,

thanks for the pull request. Looks very good overall.
I left a couple of comments with suggestions on how I would change things. Have a look.

My one big concern (it's also in the comments) is if position and tilt should be send as seperate commands. Have you checked if this is how it is done be the webcontrol server? Can they not be send in one command, e.g., '03{position}{tilt}ffff' ?

Best
Cornelius

File: warema_wms/shade.py (hunk 1 above)
Comment (thesaint1975) [reply warema_wms/shade.py]: ah, i understand, this was caused by my lack of understanding, sorry

File: warema_wms/shade.py (hunk 2 above)
Comment (thesaint1975) [reply warema_wms/shade.py]: True, overseen that

File: warema_wms/shade.py (hunk 1 above)
Comment (thesaint1975) [reply warema_wms/shade.py]: so this means if I set 100 here as default, the blinds move with this value until it is changed?

just for my unverstandig, would fit my personal needs,  could be documented so anybody else can set it regarding to his needs?
Warema default is tilt fully open, which imho is the most stupid solution for a heat and sun protection...

Comment (cornim) [reply warema_wms/shade.py]: This is only the initial value when the library (and therefore the class) is being initialized. Normally the tilt value would be passed from HA via the integration.

File: warema_wms/shade.py (hunk 5 above)
Comment (thesaint1975) [reply warema_wms/shade.py:145]: have to check this an come back

File: warema_wms/wms_controller.py (hunk 6 above)
Comment (thesaint1975) [reply warema_wms/wms_controller.py]: yes, sorry for that

File: warema_wms/wms_controller.py (hunk 7 above)
Comment (thesaint1975) [reply warema_wms/wms_controller.py:22]: agree, at least the warema app (via the webcontrol)  seems to do like that. due to the design the blinds move down fully closed, and up fully open, after they reach the position set, it takes some time (up to 10sec and more) until the set tilt value gets executed. i'm sure my approach was lead by this behavior and my lack of knowledge 

but agree, I will try to figure out how it works with just one command and come back (probably with a scream for help)

File: warema_wms/wms_controller.py (hunk 9 above)
Comment (thesaint1975) [reply warema_wms/wms_controller.py]: yep, dont have any idea what my intention was with this, sorry

File: warema_wms/wms_controller.py (hunk 8 above)
Comment (thesaint1975) [reply warema_wms/wms_controller.py:23]: i'm really not sure anymore, but i think it had worked with your command as i tried your feat_stop_branch some month ago, but I knowingly didn't update the web control box or so... 
this '01ffffffff' string i found with wireshark and as i commented, at least at my setup it works with my 6 venetian blinds (which are two different types of)

the stop command was my entrypoint into my "reverse engineering" cause it was the easiest way for me to get into it... 
actually I don't need it either, haha

File: warema_wms/wms_controller.py (hunk 10 above)
Comment (thesaint1975) [reply warema_wms/wms_controller.py]: had to google who this PEP8 is, but he's probably right :)

and thanks for the hint to this style guide, will have some looks more

File: warema_wms/wms_controller.py (hunk 8 above)
Comment (thesaint1975) [reply warema_wms/wms_controller.py:23]: would leave the decision up to you

File: warema_wms/shade.py (hunk 5 above)
Comment (thesaint1975) [reply warema_wms/shade.py:145]: technically, the angle is realized using the same pull rope as moving up and down (as mentioned elsewhere, move up = fully open, down = fully closed), i.e. in principle the status should also be true during tilt adjustment , but only for a short time (full open to fully closed needs 1 or maybe 2 seconds), no idea if that's not too short?

Comment (thesaint1975) [reply warema_wms/shade.py:145]: i did several trials yesterday, but was not able to receive a true state during tilt adjustment. 
i was not able to determine, if its caused by the by the short time it need to change from one state to another, (and i missed everytime the right time slot for the request) or the state is actually false during tilt movement.

the update command was send always via console, the tilt command via Webui of the webcontrol box. 

do you have any ideas how i could check this, pls give advice.

correction to my last post, it takes almost exactly 1 second to change tilt from full open to full closed (from the moment slats start moving)

Comment (cornim) [reply warema_wms/shade.py:145]: I would leave it as it is given that your argument that the same pull rope is used is very valid.

File: warema_wms/wms_controller.py (hunk 7 above)
Comment (cornim) [reply warema_wms/wms_controller.py:22]: If this is how it is done by the WebControl App, than it is fine like this.

File: warema_wms/wms_controller.py (hunk 8 above)
Comment (cornim) [reply warema_wms/wms_controller.py:23]: I also checked this and I did have two zeros to many in the code. Yours is correct. Thanks for catching this.

**Review by cornim — APPROVED** [review]

**Comment by cornim** [comment]
While looking through this PR I realized that there is one thing missing. So far only one type of blind was supported so it wasn't necessary to distinguish different product types but if we want to integrate this properly into HA, the library needs to carry the information which product type a shade has.
The info is returned as part of the channel request. My shades are `<produkttyp>10</produkttyp><bedientyp>4</bedientyp>`. I don't know which one is the relevant info but perhaps comparing with what you got will help us here.
Can you have a look, please?

File: warema_wms/wms_controller.py (hunk 7 above)
Comment (thesaint1975) [reply warema_wms/wms_controller.py:22]: sorry, I probably expressed myself misleadingly, I think you're right, it should only be one command.
I just didn't have it on my radar because I just "copied" the behavior of the venetian blinds as they do when there triggered via web control. via the App you can adjust tilt and position seperately, then press "play",  first the blinds move to position, after the reached its position after a view seconds the tilt is adjusted. but with only one user input.
i have to check if the App sends two commands or not (i believe not)
Unfortunately I didn't have time to look at it yesterday an check this out.
Sorry for the confusion.

Basically, my solution should be good for HA (first move into position, then set tilt), but it's not technically clean
pls correct me if i'm wrong, but i believe this is the way ho it is handled in a HA Automation (a service call for position, then a service call for tilt)

**Comment by thesaint1975** [comment]
> 

will check this hopefully tonight

File: warema_wms/wms_controller.py (hunk 7 above)
Comment (cornim) [reply warema_wms/wms_controller.py:22]: In this case you should definitely try out if this even works as one command before changing it. Could be that the webcontrol server only accepts the two commands separately (as it isn't the best implementation from what I can gather).

Comment (thesaint1975) [reply warema_wms/wms_controller.py:22]: will do my best and come back with results

**Comment by thesaint1975** [comment]
sorry for the late response,
my 6 blinds returning all the same,
`<produkttyp>0</produkttyp>
<bedientyp>2</bedientyp>`

for explanation,
4 venetian blinds are mounted in front of sloping roof windows and can take an angle of (simplified) -90° to +90° (where 0° = full open) technically speaking -90° and +90° are fully closed but with a different incidence of light depending on the position of the sun or whatever.
the other two venetian blinds are in front of the terrace doors (so they are vertically) and technically open their angle between 0° and +90°C. Inconsistently, in the scrappy Warema App these two are also displayed with a tilt adjustment range of -90° to +90°.

I'll provide exact model types later, for case it helps someone

if i can do any further check or investigation, pls let me know

regarding the move/tilt in one command, id dint have time yet, hopefully tonight

**Comment by cornim** [comment]
Thanks for looking it up. It seems that between the model that you have and the model that I have we can only distinguish between with/without tilt and we can use bedientyp for it.
If you don't have time, I can try and come up with an implementation but it might take a couple of weeks as I have very limited free time at the moment.

**Comment by thesaint1975** [comment]
unfortunately it's the same for me too (time management is very challenging with one-year-old twins, haha) 
unfortunately I'm also making no progress with move/tilt in one command and have to ask for a little patience.  
regarding the Model Type, it seems that there is only one "Funkzwischenstecker" (kind of wireless adapter) from Warema.

Regarding your HA integration, I tried to implement the stop function, added SUPPORT_STOP in the import and feature section and 
`
    def stop_cover(self, **kwargs):
        self.force_update_until = datetime.now() + timedelta(seconds=15)
        self.shade.stop_moving_shade()
`

unfortunately I always get a Shade Object has no Attribute stop_moving_shade error. do you have any idea what i missed here?
